### PR TITLE
Support rename the file has been processed.

### PR DESCRIPTION
**Pull request by casuallc**
### Motivation

Support rename the file has been processed.
#13301 

### Modifications
Add property 'processedFileSuffix'.

If set, do not delete but only rename file that has been processed.
This config only work when 'keepFile' property is false.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 
  
- [√] `no-need-doc`

### CODE REVIEW

**Comment by Anonymitaet** [comment]
@casuallc When submitting a PR, please provide doc related info [in the PR description](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation) by ticking the box or labeling a PR directly, so that Bot will recognize the info and then label the PR correctly, or else Bot can not recognize the info and then label the PR with the `doc-info-missing` label.

**Comment by Anonymitaet** [comment]
@casuallc I've labelled your PR w/ `doc`, please refer to the instructions of doc labels: https://docs.google.com/document/d/1Qw7LHQdXWBW9t2-r-A7QdFDBwmZh6ytB4guwMoXHqc0/edit#

**Comment by Haaroon** [comment]
Thanks for fixing the issue! I really appreciate your work. Renaming the file is a good idea as it keeps the operation stateless, just incase the file reader crashes and has to be restarted for some reason.

**Comment by codelipenghui** [comment]
@casuallc It's a great feature for the connector, would you please send your proposal to the dev mail list?

**Comment by casuallc** [comment]
@codelipenghui ok.

**Comment by Haaroon** [comment]
Any updates?

File: pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java [hunk 1]
```diff
@@ -112,6 +112,12 @@
      */
     private Integer numWorkers = 1;
 
+    /**
+     * If set, do not delete but only rename file that has been processed.
+     * This config only work when 'keepFile' property is false.
+     */
```
Comment (lhotari) [inline pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java:118]: Please add validation for this rule in the validate method of this class.

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java (hunk 1 above)
Comment (casuallc) [reply pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java:118]: @lhotari Fixed.

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by lhotari** [comment]
@dlg99 please review

File: pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java (hunk 1 above)
Comment (lhotari) [reply pulsar-io/file/src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java:118]: @casuallc there are some checkstyle issues:
```
Error:  src/main/java/org/apache/pulsar/io/file/FileSourceConfig.java:[174] (sizes) LineLength: Line is longer than 120 characters (found 129).
Error:  src/main/java/org/apache/pulsar/io/file/FileListingThread.java:[188] (sizes) LineLength: Line is longer than 120 characters (found 129).
Error:  Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:check (checkstyle) on project pulsar-io-file: You have 2 Checkstyle violations. -> [Help 1]
```

**Comment by codelipenghui** [comment]
@casuallc Please check the checkstyle issue.

**Review by dlg99 — APPROVED** [review]
This LGTM and a definite improvement.
There is a possible case when the file is still open for writing and is being appended to (e,g, large log files copied over a slow cross-DC network).
Another idea can include the use of the https://github.com/apache/pulsar/blob/master/pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/PulsarOffsetBackingStore.java or similar to keep name of the file and last byte offset read from it to (1) avoid reading the same content again and (2) resume reading of appended data later.

**Comment by Haaroon** [comment]
thank you

**Pull request merged**